Add support for models with mutated buffer on torch.onnx.dynamo_export

thiagocrepaldi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #114407
* __->__ #112272

This PR adds a unit test that leverages `torch.export.ExportedProgram` models that mutates registered buffers. Although the exporter already works out of the box in such scenario, the GraphModule and the exported ONNX model have extra outputs containing the mutated buffers. On future runs of the ONNX model, the mutated buffers are used as input to the model.

The aforementioned extra inputs and outputs are by design and the `ONNXProgram.model_signature` can be used to fetch detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX's, there is a mismatch between the schema because pytorch output does not include the mutated buffers present on the ONNX output.

This PR extends `onnx_program.adapt_torch_outputs_to_onnx(torch_outputs)` so that the mutated buffers are prepended to the Pytorch output, matching the ONNX schema.

Commit:
Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer

`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -424,11 +426,22 @@ def _compare_pytorch_onnx_with_ort(
     ort_outputs = run_ort(onnx_program, onnx_format_args)
 
-    if len(ref_outputs) != len(ort_outputs):
+    # When model is a torch.export.ExportedProgram, the number of outputs in the ONNX model can be greater
```
Create an output adapter step for this?

thiagocrepaldi replied:
It doesn't work, at least how we designed ` InputAdapter`  and `OutputAdapter`. They were designed to be used *after* the model is exported, to adapt input/output from pytorch to match the ONNX input/output.

The output adapter expand the pytorch output to match the ONNX output schema, so we could at most match the number of outputs. Thus, an extra output would be prepended to pytorch output to match the extra output on onnx graph.

However, the ONNX output will have an output with the *mutated* buffer while pytorch will have a not-mutated buffer as output, causing a mismatch between them.

Now if we decide to hack the purpose of the `OutputAdapter` and use it before the FX-ONNX to truncate the extra outputs, we could make it work. This is a break of contract, though.

Makes sense?

thiagocrepaldi replied:
I am creating a `adapt_onnx_outputs_to_torch` in a separate PR to help bridging this gap. Despite the output truncation, we will also need to update the mutated buffers within the model so that buffers match during tests

thiagocrepaldi replied:
done

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export using ExportedProgram with mutated buffer"


`run_test_with_fx_to_onnx_exporter_and_onnx_runtime` had to be extended
to allow extra outputs on the ONNX graph due to extra outputs due to
mutated inputs and buffers

[ghstack-poisoned]

Commit:
Update on "Support torch.onnx.dynamo_export with mutated buffers"


Due to design differences, input/output format between PyTorch model and exported ONNX model are often not the same. E.g., None is allowed for PyTorch model, but are not supported by ONNX. Nested constructs of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX, etc.

The new API replays the adapting steps recorded during export to recreate the PyTorch output format after ONNX model is executed.

Example:
```python
    >>> import torch
    >>> import torch.onnx
    >>> from typing import Dict, Tuple
    >>> def func_with_nested_input_structure(
    ...     x_dict: Dict[str, torch.Tensor],
    ...     y_tuple: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
    ... ):
    ...     if "a" in x_dict:
    ...         x = x_dict["a"]
    ...     elif "b" in x_dict:
    ...         x = x_dict["b"]
    ...     else:
    ...         x = torch.randn(3)
    ...
    ...     y1, (y2, y3) = y_tuple
    ...
    ...     return x + y1 + y2 + y3
    >>> x_dict = {"a": torch.tensor(1.)}
    >>> y_tuple = (torch.tensor(2.), (torch.tensor(3.), torch.tensor(4.)))
    >>> onnx_program = torch.onnx.dynamo_export(func_with_nested_input_structure, x_dict, y_tuple)
    >>> print(x_dict, y_tuple)
    {'a': tensor(1.)} (tensor(2.), (tensor(3.), tensor(4.)))
    >>> print(onnx_program.adapt_torch_inputs_to_onnx(x_dict, y_tuple))
    (tensor(1.), tensor(2.), tensor(3.), tensor(4.))
```

With this functionality in place, torch.onnx.dynamo_export using `ExportedProgram` as model input now supports models with mutated buffer.

When the model has mutated buffers, the ONNX model will have extra outputs, which are discarded by `ONNXProgram.adapt_onnx_outputs_to_torch`

[ghstack-poisoned]

Commit:
Update on "Support torch.onnx.dynamo_export with mutated buffers"


Due to design differences, input/output format between PyTorch model and exported ONNX model are often not the same. E.g., None is allowed for PyTorch model, but are not supported by ONNX. Nested constructs of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX, etc.

The new API replays the adapting steps recorded during export to recreate the PyTorch output format after ONNX model is executed.

Example:
```python
    >>> import torch
    >>> import torch.onnx
    >>> from typing import Dict, Tuple
    >>> def func_with_nested_input_structure(
    ...     x_dict: Dict[str, torch.Tensor],
    ...     y_tuple: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
    ... ):
    ...     if "a" in x_dict:
    ...         x = x_dict["a"]
    ...     elif "b" in x_dict:
    ...         x = x_dict["b"]
    ...     else:
    ...         x = torch.randn(3)
    ...
    ...     y1, (y2, y3) = y_tuple
    ...
    ...     return x + y1 + y2 + y3
    >>> x_dict = {"a": torch.tensor(1.)}
    >>> y_tuple = (torch.tensor(2.), (torch.tensor(3.), torch.tensor(4.)))
    >>> onnx_program = torch.onnx.dynamo_export(func_with_nested_input_structure, x_dict, y_tuple)
    >>> print(x_dict, y_tuple)
    {'a': tensor(1.)} (tensor(2.), (tensor(3.), tensor(4.)))
    >>> print(onnx_program.adapt_torch_inputs_to_onnx(x_dict, y_tuple))
    (tensor(1.), tensor(2.), tensor(3.), tensor(4.))
```

With this functionality in place, torch.onnx.dynamo_export using `ExportedProgram` as model input now supports models with mutated buffer.

When the model has mutated buffers, the ONNX model will have extra outputs, which are discarded by `ONNXProgram.adapt_onnx_outputs_to_torch`

[ghstack-poisoned]

BowenBao commented:
ci failures?

BowenBao commented on test/onnx/dynamo/test_exporter_api.py:
```diff
@@ -190,6 +190,7 @@ def test_raise_on_invalid_save_argument_type(self):
             onnx.ModelProto(),
             io_adapter.InputAdapter(),
             io_adapter.OutputAdapter(),
+            io_adapter.OutputAdapter(),
```
this looks weird

BowenBao replied:
oh i see you are adding another output adapter in this PR.

thiagocrepaldi replied:
yes, we need one output adapter to go from torch to onnx format and another to go from onnx to torch.
e.g. onnx->torch we drop extra outputs on the onnx model; torch->onnx we prepend outputs into the torch output

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -459,14 +459,24 @@ def _compare_pytorch_onnx_with_ort(
+        if output_format == "pytorch":
+            ort_outputs = onnx_program.adapt_onnx_outputs_to_torch(ort_outputs)
+        elif output_format == "onnx":
+            ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```
this is done once above on L451

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -459,14 +459,24 @@ def _compare_pytorch_onnx_with_ort(
+        "onnx",
+    ):  # Test output match with pytorch and onnx output format
+        if output_format == "pytorch":
+            ort_outputs = onnx_program.adapt_onnx_outputs_to_torch(ort_outputs)
```
I didn't do this because it felt non-trivial. onnx outputs are more organized. It can only be tensors, or (nested) list of tensors. Whereas for torch, the output can be anything. So mapping back from onnx to torch is considerably harder (how to insert back the Nones at the correct location where we discarded? how to construct back the arbitrary torch model output classes/datastructures that we flattened?).

I see that the goal was rather to eliminate the extra mutated buffers. I feel finding the way to include those outputs, when they exist because we are exporting ExportProgram, in the `ref_outputs` and have them properly accuracy checked might be more desirable.

thiagocrepaldi replied:
> I didn't do this because it felt non-trivial. onnx outputs are more organized. 
> It can only be tensors, or (nested) list of tensors. Whereas for torch, the output can be anything. So mapping back from onnx to torch is considerably harder (how to insert back the Nones at the correct location where we discarded? how to construct back the arbitrary torch model output classes/datastructures that we flattened?).

That is a good point. My PR doesn't include a test with a very complex output that need to be rebuilt. This PR only has a test which requires buffers to be prepended to the original output.

I am wondering whether `pytree` could do that for us since it records the spec of the original data?

> I see that the goal was rather to eliminate the extra mutated buffers. I feel finding the way to include those outputs, when they exist because we are exporting ExportProgram, in the `ref_outputs` and have them properly accuracy checked might be more desirable.

We already do that too :) `adapt_onnx_outputs_to_torch` is not strictly necessary to support mutated buffers, but because you had proposed in another [review](https://github.com/pytorch/pytorch/pull/109650#discussion_r1331940092) I took this opportunity to implement here and improve the UX

Maybe we can split this PR with the essentials for supporting mutated buffers and move the `adapt_onnx_outputs_to_torch` into a new change we might need to dig deeper

BowenBao commented on torch/onnx/_internal/io_adapter.py:
```diff
@@ -507,6 +515,46 @@ def apply(self, model_outputs: Any) -> Sequence[Any]:
+
+        user_outputs = []
+        for out, out_spec in zip(
+            model_outputs, self.model.graph_signature.output_specs
```
Similar concern as https://github.com/pytorch/pytorch/pull/113477#discussion_r1394972191

thiagocrepaldi replied:
resolved https://github.com/pytorch/pytorch/pull/113477#discussion_r1395014051

Commit:
Update on "Support torch.onnx.dynamo_export with mutated buffers"


Due to design differences, input/output format between PyTorch model and exported ONNX model are often not the same. E.g., None is allowed for PyTorch model, but are not supported by ONNX. Nested constructs of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX, etc.

The new API replays the adapting steps recorded during export to recreate the PyTorch output format after ONNX model is executed.

Example:
```python
    >>> import torch
    >>> import torch.onnx
    >>> from typing import Dict, Tuple
    >>> def func_with_nested_input_structure(
    ...     x_dict: Dict[str, torch.Tensor],
    ...     y_tuple: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
    ... ):
    ...     if "a" in x_dict:
    ...         x = x_dict["a"]
    ...     elif "b" in x_dict:
    ...         x = x_dict["b"]
    ...     else:
    ...         x = torch.randn(3)
    ...
    ...     y1, (y2, y3) = y_tuple
    ...
    ...     return x + y1 + y2 + y3
    >>> x_dict = {"a": torch.tensor(1.)}
    >>> y_tuple = (torch.tensor(2.), (torch.tensor(3.), torch.tensor(4.)))
    >>> onnx_program = torch.onnx.dynamo_export(func_with_nested_input_structure, x_dict, y_tuple)
    >>> print(x_dict, y_tuple)
    {'a': tensor(1.)} (tensor(2.), (tensor(3.), tensor(4.)))
    >>> print(onnx_program.adapt_torch_inputs_to_onnx(x_dict, y_tuple))
    (tensor(1.), tensor(2.), tensor(3.), tensor(4.))
```

With this functionality in place, torch.onnx.dynamo_export using `ExportedProgram` as model input now supports models with mutated buffer.

When the model has mutated buffers, the ONNX model will have extra outputs, which are discarded by `ONNXProgram.adapt_onnx_outputs_to_torch`

[ghstack-poisoned]

Commit:
Update on "Support torch.onnx.dynamo_export with mutated buffers"


Due to design differences, input/output format between PyTorch model and exported ONNX model are often not the same. E.g., None is allowed for PyTorch model, but are not supported by ONNX. Nested constructs of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX, etc.

The new API replays the adapting steps recorded during export to recreate the PyTorch output format after ONNX model is executed.

Example:
```python
    >>> import torch
    >>> import torch.onnx
    >>> from typing import Dict, Tuple
    >>> def func_with_nested_input_structure(
    ...     x_dict: Dict[str, torch.Tensor],
    ...     y_tuple: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
    ... ):
    ...     if "a" in x_dict:
    ...         x = x_dict["a"]
    ...     elif "b" in x_dict:
    ...         x = x_dict["b"]
    ...     else:
    ...         x = torch.randn(3)
    ...
    ...     y1, (y2, y3) = y_tuple
    ...
    ...     return x + y1 + y2 + y3
    >>> x_dict = {"a": torch.tensor(1.)}
    >>> y_tuple = (torch.tensor(2.), (torch.tensor(3.), torch.tensor(4.)))
    >>> onnx_program = torch.onnx.dynamo_export(func_with_nested_input_structure, x_dict, y_tuple)
    >>> print(x_dict, y_tuple)
    {'a': tensor(1.)} (tensor(2.), (tensor(3.), tensor(4.)))
    >>> print(onnx_program.adapt_torch_inputs_to_onnx(x_dict, y_tuple))
    (tensor(1.), tensor(2.), tensor(3.), tensor(4.))
```

With this functionality in place, torch.onnx.dynamo_export using `ExportedProgram` as model input now supports models with mutated buffer.

When the model has mutated buffers, the ONNX model will have extra outputs, which are discarded by `ONNXProgram.adapt_onnx_outputs_to_torch`

[ghstack-poisoned]

Commit:
Update on "Add adapt_onnx_outputs_to_torch API to torch.onnx.ONNXProgram"


Due to design differences, input/output format between PyTorch model and exported ONNX model are often not the same. E.g., None is allowed for PyTorch model, but are not supported by ONNX. Nested constructs of tensors are allowed for PyTorch model, but only flattened tensors are supported by ONNX, etc.

The new API replays the adapting steps recorded during export to recreate the PyTorch output format after ONNX model is executed.

Example:
```python
    >>> import torch
    >>> import torch.onnx
    >>> from typing import Dict, Tuple
    >>> def func_with_nested_input_structure(
    ...     x_dict: Dict[str, torch.Tensor],
    ...     y_tuple: Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]
    ... ):
    ...     if "a" in x_dict:
    ...         x = x_dict["a"]
    ...     elif "b" in x_dict:
    ...         x = x_dict["b"]
    ...     else:
    ...         x = torch.randn(3)
    ...
    ...     y1, (y2, y3) = y_tuple
    ...
    ...     return x + y1 + y2 + y3
    >>> x_dict = {"a": torch.tensor(1.)}
    >>> y_tuple = (torch.tensor(2.), (torch.tensor(3.), torch.tensor(4.)))
    >>> onnx_program = torch.onnx.dynamo_export(func_with_nested_input_structure, x_dict, y_tuple)
    >>> print(x_dict, y_tuple)
    {'a': tensor(1.)} (tensor(2.), (tensor(3.), tensor(4.)))
    >>> print(onnx_program.adapt_torch_inputs_to_onnx(x_dict, y_tuple))
    (tensor(1.), tensor(2.), tensor(3.), tensor(4.))
```

With this functionality in place, torch.onnx.dynamo_export using `ExportedProgram` as model input now supports models with mutated buffer.

When the model has mutated buffers, the ONNX model will have extra outputs, which are discarded by `ONNXProgram.adapt_onnx_outputs_to_torch`

[ghstack-poisoned]

thiagocrepaldi commented on test/onnx/onnx_test_common.py:
```diff
@@ -442,14 +442,24 @@ def _compare_pytorch_onnx_with_ort(
     ort_outputs = onnx_program(*input_args, **input_kwargs)
 
-    if len(ref_outputs) != len(ort_outputs):
+    # When model is a torch.export.ExportedProgram, the number of outputs in the ONNX model can be greater
```
https://github.com/pytorch/pytorch/pull/113686 will improve this

Commit:
Update on "Add torch.onnx.dynamo_export test for model with mutated buffer"


Add torch.onnx.dynamo_export test for model with mutated buffer

This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. The exported ONNX model has
extra outputs with the values of the mutated buffer, which can be later
used to feed the model on a next forward step.

onnx_test_common._compare_pytorch_onnx_with_ort was extended to ignore
such extra outputs (that don't exist in the original pytorch model)

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export test for model with mutated buffer"


Add torch.onnx.dynamo_export test for model with mutated buffer

This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. The exported ONNX model has
extra outputs with the values of the mutated buffer, which can be later
used to feed the model on a next forward step.

onnx_test_common._compare_pytorch_onnx_with_ort was extended to ignore
such extra outputs (that don't exist in the original pytorch model)

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export test for model with mutated buffer"


Add torch.onnx.dynamo_export test for model with mutated buffer

This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. The exported ONNX model has
extra outputs with the values of the mutated buffer, which can be later
used to feed the model on a next forward step.

onnx_test_common._compare_pytorch_onnx_with_ort was extended to ignore
such extra outputs (that don't exist in the original pytorch model)

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export test for model with mutated buffer"


Add torch.onnx.dynamo_export test for model with mutated buffer

This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. The exported ONNX model has
extra outputs with the values of the mutated buffer, which can be later
used to feed the model on a next forward step.

onnx_test_common._compare_pytorch_onnx_with_ort was extended to ignore
such extra outputs (that don't exist in the original pytorch model)

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export test for model with mutated buffer"


Add torch.onnx.dynamo_export test for model with mutated buffer

This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. The exported ONNX model has
extra outputs with the values of the mutated buffer, which can be later
used to feed the model on a next forward step.

onnx_test_common._compare_pytorch_onnx_with_ort was extended to ignore
such extra outputs (that don't exist in the original pytorch model)

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export test for model with mutated buffer"


Add torch.onnx.dynamo_export test for model with mutated buffer

This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. The exported ONNX model has
extra outputs with the values of the mutated buffer, which can be later
used to feed the model on a next forward step.

onnx_test_common._compare_pytorch_onnx_with_ort was extended to ignore
such extra outputs (that don't exist in the original pytorch model)

[ghstack-poisoned]

Commit:
Update on "Add torch.onnx.dynamo_export test for model with mutated buffer"


Add torch.onnx.dynamo_export test for model with mutated buffer

This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. The exported ONNX model has
extra outputs with the values of the mutated buffer, which can be later
used to feed the model on a next forward step.

onnx_test_common._compare_pytorch_onnx_with_ort was extended to ignore
such extra outputs (that don't exist in the original pytorch model)

[ghstack-poisoned]

Commit:
Update on "Add support for models with mutated buffer on torch.onnx.dynamo_export"


This PR adds a unit test that leverages torch.export.ExportedProgram
models that mutates registered buffers. Although the exporter works out
of the box in such scenario, the GraphModule and the exported ONNX model
have extra outputs with the values of the mutated buffer.
On future forward calls, the mutated buffers are used as input to the
model.

This is all by design and the `ONNXProgram.model_signature` can be used
to have a detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX output, there is
a mismatch between the schema because pytorch output does not include
the mutated buffers.

This PR updates
onnx_program.adapt_torch_outputs_to_onnx(pytorch_outputs) do that the
mutated buffers are prepended to the output, matching the ONNX schema

[ghstack-poisoned]

Commit:
Update on "Add support for models with mutated buffer on torch.onnx.dynamo_export"


This PR adds a unit test that leverages `torch.export.ExportedProgram` models that mutate registered buffers.
Although the exporter already works out of the box in such scenario, the `GraphModule` and the exported ONNX model have extra outputs with the values of the mutated buffer. On future forward calls, the mutated buffers are used as input to the
model. This is all by design and the `ONNXProgram.model_signature` can be used to have a detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX output, there is a mismatch between the schema because pytorch output does not include the mutated buffers.

This PR updates onnx_program.adapt_torch_outputs_to_onnx(pytorch_outputs) do that the mutated buffers are prepended to the output, matching the ONNX schema.

`ONNXProgram` now have a reference (not copy) of the original `torch.export.ExportedProgram`, this is needed so that `ONNXProgram` can have access to the user model `state_dict` which contains the buffers, used on `adapt_torch_outputs_to_onnx` calls

[ghstack-poisoned]

Commit:
Update on "Add support for models with mutated buffer on torch.onnx.dynamo_export"


This PR adds a unit test that leverages `torch.export.ExportedProgram` models that mutate registered buffers.
Although the exporter already works out of the box in such scenario, the `GraphModule` and the exported ONNX model have extra outputs with the values of the mutated buffer. On future forward calls, the mutated buffers are used as input to the
model. This is all by design and the `ONNXProgram.model_signature` can be used to have a detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX output, there is a mismatch between the schema because pytorch output does not include the mutated buffers.

This PR updates onnx_program.adapt_torch_outputs_to_onnx(pytorch_outputs) do that the mutated buffers are prepended to the output, matching the ONNX schema.

`ONNXProgram` now have a reference (not copy) of the original `torch.export.ExportedProgram`, this is needed so that `ONNXProgram` can have access to the user model `state_dict` which contains the buffers, used on `adapt_torch_outputs_to_onnx` calls

[ghstack-poisoned]

Commit:
Update on "Add support for models with mutated buffer on torch.onnx.dynamo_export"


This PR adds a unit test that leverages `torch.export.ExportedProgram` models that mutates registered buffers. Although the exporter already works out of the box in such scenario, the GraphModule and the exported ONNX model have extra outputs containing the mutated buffers. On future runs of the ONNX model, the mutated buffers are used as input to the model.

The aforementioned extra inputs and outputs are by design and the `ONNXProgram.model_signature` can be used to fetch detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX's, there is a mismatch between the schema because pytorch output does not include the mutated buffers present on the ONNX output.

This PR extends `onnx_program.adapt_torch_outputs_to_onnx(torch_outputs)` so that the mutated buffers are prepended to the Pytorch output, matching the ONNX schema.

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -942,6 +942,32 @@ def forward(self, x):
         )
 
+    @pytorch_test_common.xfail_if_model_type_is_not_exportedprogram(
+        "Unsupported FX nodes: {'call_function': ['aten.add_.Tensor']}. "
```
@BowenBao should Functionalization take care of this?

titaiwangms replied:
Looks like discrepency between our Functionalization and the one inside torch.export.export?

titaiwangms replied:
Make sure we have an issue tracking this.

thiagocrepaldi replied:
https://github.com/pytorch/pytorch/issues/114406

BowenBao replied:
yea because dynamo_export was not lifting buffers, there was another issue you can link it there too.

titaiwangms commented on test/onnx/onnx_test_common.py:
```diff
@@ -436,16 +436,18 @@ def _compare_pytorch_onnx_with_ort(
+    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
     ort_outputs = onnx_program(*input_args, **input_kwargs)
+    ref_outputs = ref_model(*ref_input_args, **ref_input_kwargs)
+    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```
Why wasn't this a requirement? Could you elaborate more?

thiagocrepaldi replied:
there 2 changes here.

one is just aesthetic, not thing really changed. 
Before we had `    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_model(*ref_input_args,**ref_input_kwargs))` and now we have 
```python
    ref_outputs = ref_model(*ref_input_args, **ref_input_kwargs)
    ref_outputs = onnx_program.adapt_torch_outputs_to_onnx(ref_outputs)
```

The second change is swapping the order between `ref_outputs` and `ort_outputs` which is explained by the new comment:
```python
    # ONNXProgram holds a reference (not copy) to the original ref_model, including its state_dict.
    # Thus, ONNXProgram() must run before ref_model() to prevent ref_model.forward() from changing the state_dict.
    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
```

titaiwangms replied:
My question was why we didn't need the second change? Because having the need of the second change doesn't feel very intuitive.

thiagocrepaldi replied:
Offline discussion. When the input model is a `torch.nn.Module`, the FX tracer cannot handle buffer mutation, so the order in which ORT model and Torch model were never an issue. When the input model is `torch.export.ExportedProgram`, we internally use AOTAutograd which lifts mutated buffers as model inputs/outputs, so we need to execute ORT (with the pytorch model state) before torch to make sure both models are on the same iteration (forward step).

titaiwangms reviewed (approved):
Please make a NOTE saying mutated buffers are no longer static after torch.export.export, and test if non-registered buffer is also dynamic. Thanks!

Commit:
Update on "Add support for models with mutated buffer on torch.onnx.dynamo_export"


This PR adds a unit test that leverages `torch.export.ExportedProgram` models that mutates registered buffers. Although the exporter already works out of the box in such scenario, the GraphModule and the exported ONNX model have extra outputs containing the mutated buffers. On future runs of the ONNX model, the mutated buffers are used as input to the model.

The aforementioned extra inputs and outputs are by design and the `ONNXProgram.model_signature` can be used to fetch detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX's, there is a mismatch between the schema because pytorch output does not include the mutated buffers present on the ONNX output.

This PR extends `onnx_program.adapt_torch_outputs_to_onnx(torch_outputs)` so that the mutated buffers are prepended to the Pytorch output, matching the ONNX schema.

[ghstack-poisoned]

Commit:
Update on "Add support for models with mutated buffer on torch.onnx.dynamo_export"


This PR adds a unit test that leverages `torch.export.ExportedProgram` models that mutates registered buffers. Although the exporter already works out of the box in such scenario, the GraphModule and the exported ONNX model have extra outputs containing the mutated buffers. On future runs of the ONNX model, the mutated buffers are used as input to the model.

The aforementioned extra inputs and outputs are by design and the `ONNXProgram.model_signature` can be used to fetch detailed input/output schema for the exported model.

However, when we want to compare pytorch output to ONNX's, there is a mismatch between the schema because pytorch output does not include the mutated buffers present on the ONNX output.

This PR extends `onnx_program.adapt_torch_outputs_to_onnx(torch_outputs)` so that the mutated buffers are prepended to the Pytorch output, matching the ONNX schema.

[ghstack-poisoned]

thiagocrepaldi commented:
@pytorchbot merge

BowenBao commented on test/onnx/onnx_test_common.py:
```diff
@@ -436,16 +436,18 @@ def _compare_pytorch_onnx_with_ort(
-
+    # ONNXProgram holds a reference (not copy) to the original ref_model, including its state_dict.
+    # Thus, ONNXProgram() must run before ref_model() to prevent ref_model.forward() from changing the state_dict.
+    # Otherwise, the ref_model can change buffers on state_dict which would be used by ONNXProgram.__call__()
```
I recall there was a PR that will make ONNXProgram.__call__() update original pytorch model lol. This problem will arise again if that's added.

Btw there is a flag `has_mutation` up there that triggers model cloning. Might be helpful.

thiagocrepaldi replied:
that pr was merged with this one.

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting you change but it is failing dynamo test in trunk https://hud.pytorch.org/pytorch/pytorch/commit/c4a22d6918b7ca218f2712d7e7e147aca7127fa3' -c weird

huydhn commented:
It looks like the revert doesn't help, I will reland the change.  Sorry for the churn.

huydhn commented:
@pytorchbot merge
